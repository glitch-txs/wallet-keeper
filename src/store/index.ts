import { create } from 'zustand'
import { persist } from 'zustand/middleware'
import type { WalletStore } from '../types'
import { mainnet, type Chain } from 'viem/chains'
import type { Address } from 'viem'
import { getTokenProperties } from '../utils/token'
import { generateWallet } from '../utils/wallet'

export const useWalletStore = create<WalletStore>()(
	persist(
		(set, get) => ({
			wallets: [],
			activeWallet: undefined,
			activeChain: mainnet,
			tokens: [],

			addWallet: ({ name, password }: { name: string; password: string }) => {
				const newWallet = generateWallet({ name, password })
				set((state) => ({
					wallets: [...state.wallets, newWallet],
				}))
				set({ activeWallet: newWallet })
			},

			removeWallet: (address: string) => {
				set((state) => ({
					wallets: state.wallets.filter((w) => w.address !== address),
					activeWallet:
						state.activeWallet?.address === address
							? state.wallets.filter((w) => w.address !== address)[0]
							: state.activeWallet,
				}))
			},

			setActiveWallet: (address: string) => {
				const wallet = get().wallets.find((w) => w.address === address)
				if (wallet) {
					set({ activeWallet: wallet })
				}
			},

			addToken: async({ address, chain }: { address: Address, chain: Chain }) => {
				const newToken = await getTokenProperties({ address, chain })
				set((state) => ({
					tokens: [...state.tokens, newToken],
				}))
			},

			removeToken: (address: string) => {
				set((state) => ({
					tokens: state.tokens.filter((t) => t.address !== address),
				}))
			},
		}),
		{
			name: 'wallet-storage',
			partialize: (state) => ({ wallets: state.wallets, activeWallet: state.activeWallet }),
		},
	),
)
