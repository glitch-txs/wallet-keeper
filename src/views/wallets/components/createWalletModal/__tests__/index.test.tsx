import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import toast from 'react-hot-toast'
import { useWalletStore } from '../../../../../store'
import CreateWalletModal from '..'

vi.mock('react-hot-toast', () => ({
  default: {
    success: vi.fn(),
    error: vi.fn(),
  }
}))

const mockAddWallet = vi.fn()
const mockSetIsModalOpen = vi.fn()

beforeEach(() => {
  vi.clearAllMocks()

  useWalletStore.setState((state) => ({
    ...state,
    addWallet: mockAddWallet,
  }))
})

describe('CreateWalletModal Component', () => {
  it('renders modal correctly when open', () => {
    render(<CreateWalletModal isModalOpen={true} setIsModalOpen={mockSetIsModalOpen} />)

    expect(screen.getByText('Create a New Wallet')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Enter wallet name')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Enter password')).toBeInTheDocument()
  })

  it('shows validation errors when fields are empty', () => {
    render(<CreateWalletModal isModalOpen={true} setIsModalOpen={mockSetIsModalOpen} />)

    fireEvent.click(screen.getByText('Create Wallet'))

    expect(screen.getByText('Wallet Name is required.')).toBeInTheDocument()
    expect(screen.getByText('Password is required.')).toBeInTheDocument()
  })

  it('calls addWallet and closes modal on successful creation', () => {
    mockAddWallet.mockReturnValue({ name: 'Test Wallet', password: 'testpass' })

    render(<CreateWalletModal isModalOpen={true} setIsModalOpen={mockSetIsModalOpen} />)

    fireEvent.change(screen.getByPlaceholderText('Enter wallet name'), { target: { value: 'Test Wallet' } })
    fireEvent.change(screen.getByPlaceholderText('Enter password'), { target: { value: 'testpass' } })
    fireEvent.click(screen.getByText('Create Wallet'))

    expect(mockAddWallet).toHaveBeenCalledWith({ name: 'Test Wallet', password: 'testpass' })
    
    expect(toast.success).toHaveBeenCalledWith('Wallet Created Successfully!')
    
    expect(mockSetIsModalOpen).toHaveBeenCalledWith(false)

    expect(screen.getByPlaceholderText('Enter wallet name')).toHaveValue('')
    expect(screen.getByPlaceholderText('Enter password')).toHaveValue('')
  })

  it('shows error toast if addWallet fails', () => {
    mockAddWallet.mockReturnValue(null)

    render(<CreateWalletModal isModalOpen={true} setIsModalOpen={mockSetIsModalOpen} />)

    fireEvent.change(screen.getByPlaceholderText('Enter wallet name'), { target: { value: 'Test Wallet' } })
    fireEvent.change(screen.getByPlaceholderText('Enter password'), { target: { value: 'testpass' } })
    fireEvent.click(screen.getByText('Create Wallet'))

    expect(toast.error).toHaveBeenCalledWith('An error occurred.')
  })

  it('matches snapshot', () => {
    const { container } = render(<CreateWalletModal isModalOpen={true} setIsModalOpen={mockSetIsModalOpen} />)
    expect(container).toMatchSnapshot()
  })
})
