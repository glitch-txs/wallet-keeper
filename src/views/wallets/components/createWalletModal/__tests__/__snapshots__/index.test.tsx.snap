// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateWalletModal Component > matches snapshot 1`] = `<div />`;
